provider/datagen/src/provider.rs

```diff
@@ -485,3 +510,34 @@ impl SourceData {
             .locales(levels.iter().copied())
     }
 }
+
+pub(crate) trait IterableDataProviderInternal<M: KeyedDataMarker>: DataProvider<M> {
+    fn supported_locales_impl(&self) -> Result<Vec<DataLocale>, DataError>;
+}
+
+impl<M: KeyedDataMarker> IterableDataProvider<M> for DatagenProvider
+where
+    DatagenProvider: IterableDataProviderInternal<M>,
+{
+    fn supported_locales(&self) -> Result<Vec<DataLocale>, DataError> {
+        #[allow(deprecated)] // SourceData
+        self.source
+            .supported_locales_cache_vec
+            .0
+            .insert_with(M::KEY, || Box::from(self.supported_locales_impl()))
+            .as_ref()
+            .cloned()
+            .map_err(|e| *e)
+    }
+
+    fn supports_locale(&self, locale: &DataLocale) -> Result<bool, DataError> {
```

Comment:
`supports_locale` is called n times per key, whereas `supported_locales` is only called once. It probably makes more sense to cache a `HashSet` and convert it to a `Vec` for the one call, than to linearly scan a `Vec` n times.

Reply:
Performance on `cargo make bakeddata datetime` (run time only, not build time):

| Commit | Trial 1 | Trial 2 | Trial 3 |
|---|---|---|---|
| 11a1ed1 (Vec) | 53.18 | 61.74 | 53.82 |
| f2de286 (HashSet) | 58.64 | 51.89 | 51.03 |

So it looks like HashSet wins.